Docs: rewrite mobile OTA and Dev Tool documentation (#9061)
Couchers-org/couchers: docs/native-ota-updates.md

# Mobile app Over-the-Air (OTA) updates

## Why OTA and what it does.

Over-the-Air (OTA) updates are a way to update most of the mobile app by shipping a new bundle of code to the native app, without going through the app store.

> We have traditionally controlled our entire stack: when users visit the website, we can at any time replace the backend and frontend code in order to fix bugs or add new features. We can go from identifying a bug to having it fixed in minutes. We are very comfortable with this and it allows us to also be a bit aggressive (aka "test in prod").

Reviewer: I don't think we should encourage to test in prod.

IMO it's more clear to say it allows mobile development to be more like web development, rather than waiting for the next build to be approved and users to update the app. This means we can quickly fix most bugs, etc.

Author: fixed

We use React Native through the Expo framework. React Native basically uses JavaScript to manipulate a virtual DOM like normal react, but instead of web React, you don't manipulate `<div>`s, etc instead you manipulate `<ThemedText>`s and `<View>`s that map to actual native iOS/android components.

This means we have three layers:

1. The actual "native base/module": Objective-C (iOS) or Java (Android) code that runs a JavaScript engine (called Hermes) and does things like goes from a `<ThemedText>` cross-platform abstraction to true native components in an appropriate UI framework on the given mobile operating system. This also includes things like providing a unified API to push notifications, and other Expo modules we pull in for common use cases. By and large we do not have to touch this, other than pulling in existing dependencies or tiny bits of code. Expo provides most of this.
2. Our "native shell" code: the JavaScript code that defines where WebViews are, where ThemedText goes, etc. This is interpreted by the native base layer via Hermes and manipulates that code.
3. Our web app: this is loaded within the WebViews from our servers every time the user uses the platform.

> We have full control at all times over (3), and OTA updates give us relatively flexible control over (2). We cannot change (1) without going through the app stores. But this pyramid is inverted when it comes to frequency of changes: we only need to touch (1) in very rare cases where we either fully upgrade Expo, or where we add a new native Expo module (like geolocation, etc). We'll sometimes want to touch (2) or if we start moving towards native screens, we'll want to do it with higher frequency. More than half the time though, our changes touch (3), as this is the shared layer used by all clients across web and the apps. This means that OTA updates give us the ability to do something north of 95% of our work without going through the app stores.

Reviewer: This is a lot of text, I wonder if a bit of an architecture diagram of the layers would make it more understandable at a glance?

Author: i am not sure how that would look?

Reviewer: It's just a lot of text.

Author: ok lemme try

Author: added a diagram!

### On compatibility: fingerprints and runtimeVersion

> It would be a beautiful world if Expo provided a stable ABI between the native base and the native shell (parts (1) and (2) above). However, this is not the case. You need to be very careful to make sure that you only ship code to clients that actually matches the app downloaded from the app stores.

Reviewer: ```suggestion
It would be a beautiful world if Expo provided a stable API between the native base and the native shell (parts (1) and (2) above). However, this is not the case. You need to be very careful to make sure that you only ship code to clients that actually matches the app downloaded from the app stores.
```

Author: i do mean the Application Binary Interface here

In order to do this, Expo has a "runtimeVersion", that it compares before applying an OTA update to the running app. By default, this runtime version is a hash of a bunch of different data, that is supposed to catch you making breaking changes. This is called a fingerprint. You can actually just make it whatever, if you want to YOLO it, and I believe Expo has a rollback mechanism so it won't brick the app. Still, we need to be very careful with only shipping OTA updates with code that is compatible with the native base.

> 

Reviewer: I don't think we need the "you can actually just make it whatever YOLO it" part.

Let's try and keep it on a need to know basis as it's a lot to read.

Author: got rid of the yolo part

> Currently we use fingerprints as our runtime version. You can decide what stuff to include in the fingerprint (which information to watch for changes with). We use some tiny bit unsafe options, which *may* cause OTA update issues. This is so that we have some more freedom to make safe changes. But in order to catch any potential issues, we also have a "full" fingerprint that we track, which includes everything.

Reviewer: What are the "tiny bit unsafe options"?

Author: i'm not sure of all the sharp edges, that's why we have the full fingerprints for safety. e.g. right now we don't include all the package scripts in the fingerprint. technically you can do arbitrary things with the package scripts that would break compatibility; but almost never do you expect us to do that. hence the safety mechanism

Reviewer: can you update the text to indicate what you wrote here?

Author: yep; updated

So basically, fingerprints tell you: "here's a unique value that captures all the stuff about the native code that may break OTA updates".

Don't change/regenerate the fingerprints without consulting mobile dev leads. Even the pedantic full ones.

When fingerprints change, we must publish a new version via app stores, and users have to download it. We cannot ship OTA updates to old clients.

## How OTA updates work

### On the client

> I keep saying OTA updates give us "flexibility", but not full control.

Reviewer: ```suggestion
OTA updates give us "flexibility", but not full control.
```

### Finding out about a new update

In order to find out about a new update, Expo uses the "Expo Updates Protocol v1". This is a relatively straightforward protocol that includes some special headers, and where the backend basically tells the client one of three things: (1) there's no update for you right now, (2) there's a new bundle, and here's where you can find it, or (3) sorry we messed it up, revert to whatever code you shipped with from the app store.

### Our backend implementation

Our backend serves the update protocol in `Bugs.GetNativeUpdateManifest`, which the client loads from `https://api.couchers.org/native/ota/manifest`. This endpoint decides which update to serve, depending on various logic.

> The backend generally picks the newest non-banned OTA update for the combination of platform and fingerprint available from teh database, then serves pre-signed manifest bytes that are fetched from the CDN. (Banning is a mechanism to quickly stop us from distributing botched updates.)

Reviewer: ```suggestion
The backend generally picks the newest non-banned OTA update for the combination of platform and fingerprint available from the database, then serves pre-signed manifest bytes that are fetched from the CDN. (Banning is a mechanism to quickly stop us from distributing botched updates.)
```

### Publishing new updates

> The publishing flow involes basically taking a full Expo bundle (generated via `ota-bundle.mjs`), stamping it with a new publish timestamp, signing it, producing a manifest file, and telling the backend about it.

Reviewer: ```suggestion
The publishing flow involves taking a full Expo bundle (generated via `ota-bundle.mjs`), stamping it with a new publish timestamp, signing it, producing a manifest file, and telling the backend about it.
```

> You can see how this works by tracing the staging OTA update flow, where we build a new update on every push to `develop`. For prod the flow is the same, but we have a magic button in a magic place to trigger it.

Reviewer: Where do we trace the staging OTA update flow?

Author: from .gitlab-ci.yml; added

> You can see how this works by tracing the staging OTA update flow, where we build a new update on every push to `develop`. For prod the flow is the same, but we have a magic button in a magic place to trigger it.

Reviewer: "A magic button in a magic place" is pretty vague, can you make this more clear? It's the Ops dashboard right? Can we just say a button the ops dashboard that only leads have access to?

Author: yeah it's the ops dash, but i'm not sure we need to mention that; given the ops page may change.

> We can "rollback" by taking an old bundle, restamp & resign it, etc; and it is eaten up by the clients which are no wiser about the fact that it's actually older.

Reviewer: Should we define the rollback process somewhere?

Author: it is exactly the same thing as deploying a new bundle, it just re-publishes an old one; so there is nothing new

Reviewer: So it's just a revert PR? We upload the bundle somewhere? I don't get it. Can we make that really clear as usually it's going to be a bit of a crisis when it's happening no?

Author: good point; i added a concrete step on how to do it
